packages/pluginmanager/src/model.ts

```diff
@@ -0,0 +1,474 @@
+    this._pluginData = options.pluginData;
+    this._serverSettings =
+      options.serverSettings || ServerConnection.makeSettings();
+    this._query = options.query || '';
+    this._isDisclaimed = options.isDisclaimed ?? false;
+    this._extraLockedPlugins = options.extraLockedPlugins ?? [];
+    this.refresh()
+      .then(() => this._ready.resolve())
+      .catch(e => this._ready.reject(e));
+    this._trans = (options.translator ?? nullTranslator).load('jupyterlab');
+  }
+
+  get available(): ReadonlyArray<IEntry> {
+    return [...this._available.values()];
+  }
+
+  /**
+   * Contains an error message if an error occurred when querying plugin status.
+   */
+  statusError: string | null = null;
+
+  get isLoading(): boolean {
+    return this._isLoading;
+  }
+
+  /**
+   * Enable a plugin.
+   *
+   * @param entry An entry indicating which plugin to enable.
+   */
+  async enable(entry: IEntry): Promise<void> {
+    if (!this.isDisclaimed) {
+      throw new Error('User has not confirmed the dislaimer');
+    }
+    await this._performAction('enable', entry);
+    entry.enabled = true;
+  }
+
+  /**
+   * Disable a plugin.
+   *
+   * @param entry An entry indicating which plugin to disable.
+   * @returns Whether the plugin was disabled
+   */
+  async disable(entry: IEntry): Promise<void> {
+    if (!this.isDisclaimed) {
+      throw new Error('User has not confirmed the dislaimer');
+    }
+    const { dependants, optionalDependants } = this.getDependants(entry);
+    if (dependants.length > 0) {
+      // We require user to disable plugins one-by-one as each of them may have
+      // further dependencies (or optional dependencies) and we want the user to
+      // take a pause to think about those.
+      void showDialog({
+        title: this._trans.__('This plugin is required by other plugins'),
+        body: PluginRequiredMessage({
+          plugin: entry,
+          dependants,
+          trans: this._trans
+        }),
+        buttons: [Dialog.okButton({ label: this._trans.__('Ok') })]
+      });
+      return;
+    }
+    if (optionalDependants.length > 0) {
+      const userConfirmation = await showDialog({
+        title: this._trans.__('This plugin is used by other plugins'),
+        body: PluginInUseMessage({
+          plugin: entry,
+          optionalDependants,
+          trans: this._trans
+        }),
+        buttons: [
+          Dialog.okButton({ label: this._trans.__('Disable anyway') }),
+          Dialog.cancelButton({ label: this._trans.__('Cancel') })
+        ]
+      });
+      if (!userConfirmation.button.accept) {
+        return;
+      }
+    }
+
+    await this._performAction('disable', entry);
+    if (this.actionError) {
+      return;
+    }
+    entry.enabled = false;
+  }
+
+  protected getDependants(entry: IEntry): {
+    dependants: IEntry[];
+    optionalDependants: IEntry[];
+  } {
+    const dependants = [];
+    const optionalDependants = [];
+    if (entry.provides) {
```

Comment:
Done, but decided to keep it as object with `name` and `description` properties (so that we can expose more in future if needed), so this fragment is without change. The change was in b884d27e911359345b04fe6b9a44aa5fabd3923e.